### PR TITLE
Basic logging support

### PR DESCRIPTION
**Pull request by richfitz**
This PR implements logging for outpack. I had used the lgr package, which we've used elsewhere but it was proving very slow to clean up (I believe this was tidying up the globally stored logger objects, but took several seconds of unexplained hang at the end of the session!), and it was a bit of a fight in the end anyway (see code around 0dbe283 if curious). Some vestiges of this remain in the design.

Overall the idea is somewhat log4j-ish in that we have log levels, though we only support info/debug/trace in practice. Logging defaults can be configured on the root - in particular the default log threshold and whether we print to the console. Individual packets can override this at creation, and packets also include a json log target, so that each packet creation ends up with a `log.json` in it containing the output from running a script and all the timings of various outpack commands.

There's a lot to go at here because of the configuration, implementation and various bits of fighting with printing logs as expected, but I think it hangs together ok.

Things that are not dealt with yet - I don't think that one `outpack_packet_run` can now successfully call another; some care will be needed to deal with the way that output stream nest.  Something to think about later, as this is currently not at all a possibility in orderly. There's some notes of pain in updating configuration, too. I've not added all possible places where we might want logging (e.g., when rebuilding an index we might want some trace-level logging) as I figured we can make those changes as we need them. This PR does include most of the useful logging for orderly though.

Note that `outpack_packet_run` has lost its `echo` argument, with that moving into the logging control

See https://github.com/mrc-ide/orderly3/pull/12 for changes required in orderly3

### CODE REVIEW

File: R/packet.R [hunk 1]
```diff
@@ -391,6 +422,7 @@ outpack_packet_clear <- function() {
 
 outpack_packet_finish <- function(packet) {
   packet$complete <- TRUE
+  unlink(packet$logger$appenders$json$destination)
```
Comment (richfitz) [inline R/packet.R]: this line needs removing, left over from the lgr bits...

File: R/logging.R [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    ret$threshold <- log_level_check(threshold)
+  }
+
+  ret
+}
+
+
+log_console <- function(topic, detail, caller, log_level) {
+  ## Filter out some log types; these will never want echoing to the
+  ## console like via this function and I imagine that this list will
```
Comment (hillalex) [inline R/logging.R]:
```suggestion
  ## console via this function and I imagine that this list will
```

File: R/packet.R [hunk 3]
```diff
@@ -30,6 +30,17 @@
...
+##' @param logging_console Optional logical, indicating if we should
+##'   override the root's default in logging to the console. A value
+##'   of `NULL` uses the root value, `TRUE` enables console output
+##'   even when this is suppressed by the root, and `FALSE` disables
+##'   it even when this is enabled by the root.
+##'
+##' @param logging_threshold Optional log threshold, indicating if we
+##'   override the root's default in logging to the console. A value
+##'   of `NULL` uses the root value, otherwise use a `lgr`-compatible
+##'   log threshold (e.g., `debug`, `all`, etc)
```
Comment (hillalex) [inline R/packet.R]: this is no longer true right, can only be one of `info`, `debug`, `trace`? similarly in other roxygen comments for `logging_threshold`

**Review by hillalex — APPROVED** [review]
This works super nicely and the interface is great!

> I don't think that one outpack_packet_run can now successfully call another;

I had no idea this was an option but would be really interesting for orchestrating workflows, then a workflow could just be another packet right? Although I guess in practice might not be super useful as would make it harder to debug/rerun steps.

File: R/packet.R (hunk 3 above)
Comment (richfitz) [reply R/packet.R]: you're right, thanks

**Comment by richfitz** [comment]
> I had no idea this was an option but would be really interesting for orchestrating workflows, then a workflow could just be another packet right? Although I guess in practice might not be super useful as would make it harder to debug/rerun steps.

it would still be within one packet, so no use for orchestrating workflows, but would allow one script to call another and so on. probably what we need to do in the short term is just prevent recursive calling, which would be easy enough I think.

**Comment by hillalex** [comment]
> it would still be within one packet, so no use for orchestrating workflows, but would allow one script to call another and so on. probably what we need to do in the short term is just prevent recursive calling, which would be easy enough I think.

Ah I see

**Review by r-ash — APPROVED** [review]
This all looks great

**Pull request merged**